airflow/jobs/scheduler_job_runner.py

```diff
@@ -1838,12 +1890,57 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
     def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[TaskInstance]]:
         """Organize TIs into lists per their respective executor."""
         _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
-        executor: str | None
         for ti in tis:
-            if ti.executor:
-                executor = str(ti.executor)
-            else:
-                executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                _executor_to_tis[executor_obj].append(ti)
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            if TYPE_CHECKING:
+                # All executors should have a name if they are initted from the executor_loader. But we need
+                # to check for None to make mypy happy.
+                assert executor.name
+            executor_to_slots_available[executor.name] = executor.slots_available
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                if TYPE_CHECKING:
+                    # All executors should have a name if they are initted from the executor_loader. But we
+                    # need to check for None to make mypy happy.
+                    assert executor_obj.name
+                if executor_to_slots_available[executor_obj.name] > 0:
```

Comment:
so something seems weird here.  your code assumes that `executor_to_slots_available` already has executor in there.  but from looking at the code it seems possible that `tis` might have an executor that hasn't been loaded yet.  othrewise why do we have to "try to load it" every time?  so then logically it must be possible that `executor_to_slots_available` might not have this executor in it. so then this could would fail when trying to access the executor.  you might want to fix that.  as with all of these suggestions, i'm not trying to block your PR (that's why i don't do "request changes").  but you may may want to address that either here or in a followup.

Reply:
Hey no worries!

This is what we've talked about a couple times now. The DAG parser will catch any TI that is trying to use an executor that isn't configured.
We load the executor here since the TI only has a string representation of the executor, which could be a module path, alias, class name. So we need to pass that to the executor loader to get back an executor object and we use the `name` from then on canonically inside the scheduler (because a dict with all those string keys would be wild and hard to reason about). The executor loader caches all the executors and their names so it's an O(1) lookup to load the executor given the string name.

I hope that clarifies things! If you still see an issue do let me know!